File: chapter3/chapter3.py
import math
import copy
import random

class Arrays():
  def __init__(self):
    self.array = []
    self.check = lambda arr, pos: pos < 0 or pos >= len(self.array) 

  def push_front(self, value):  
    self.array = [value] + self.array
    self.insert_at(0, value)
    return self

  def pop_front(self):
    return self.remove_at(0);
  
  def insert_at(self, pos, val):
    if not self.check(self.array, pos):
      self.array[pos] = val
    return self 

  def remove_at(self, pos):
    value = None
    if self.check(self.array, pos):
      return None
    value = self.array[pos]
    self.array = self.array[0:pos] + self.array[pos + 1: :] 
    return value

  def swap_pairs(self):
    for i in range (0, int(math.floor( len(self.array) / 2))):
      self.array[i], self.array[len(self.array) - 1 -  i] = self.array[len(self.array) - 1 - i], self.array[i]  
  
  def remove_dup_order(self):
    index = 0
    dupping_mode = False
    while index < len(self.array):
      if self.array[index - 1] == self.array[index]:
        self.remove_at(index)
      else:
        index += 1 
  
  def min_to_front(self):
    obj = [None, None]
    for i in range(0, len(self.array)):
      if obj[0] == None or self.array[i] < obj[0]:
        obj[0] = self.array[i]
        obj[1] = i
    if obj[0]:
      self.push_front( self.remove_at(obj[1]))
  
  def reverse(self):
    index = 0
    mid =  int (math.floor( len(self.array)/2))
    size = len(self.array)
    for i in range (0, mid):
      self.array[i],  self.array[ size - i - 1] = self.array[size - i - 1], self.array[i]   
    return self.array

  def rotate(self, shift_by):
    arr = self.array
    left_rot_en = shift_by < 0
    
    if left_rot_en:
      shift_by *= -1

    for i in range(0, shift_by):
      if shift_by <  0:
        for k in range(0,len(arr) - 1):
          arr[k + 1], arr[k] = arr[k] , arr[k + 1] 
      else:
        for k in range( len(self.array) - 2 ,-1, -1 ):
          arr[k + 1], arr[k] = arr[k] , arr[k + 1] 
  
  def filter_range(self, min, max):
    result = []
    arr =  self.array  
    for i, c in enumerate(self.array):
      if  c >min and c < max:
        result.append(c)    
    return result

  def concat(self,a,b):
    result = []
    index_a = 0
    index_b = 0

    while index_a< len(a)  or  index_b < len(b):

      if index_a < len(a):
        result.append(a[index_a])
        index_a+= 1

      elif index_b < len(b):
        result.append(b[index_b])
        index_b += 1
    return result
     
  def remove_negative(self):
    index = 0

    while index < len(self.array):
      if self.array[index] < 0 :
        self.remove_at(index)
      else :
        index += 1

  def second_to_last(self):
    count = 0
    for i in range(len(self.array) - 1, -1, -1):
      count += 1
      if count == 2:
        return self.array[i]
    return None


  def second_largest(self):
    buffer = [None, None]
    for i in range(0, len(self.array)):
      data = self.array[i]
      if i < 2:
        buffer[i] = data
      elif data > buffer[0] or data > buffer[1]:
        if buffer[0] > buffer[1]:
          buffer[0], buffer[1] = buffer[1], buffer[0]
        if data > buffer[0]:
          buffer[1] = buffer[0] 
          buffer[0] = data         
        elif data > buffer[1]:
          buffer[1]  = data 
    return buffer[0]       
      
  def nth_to_last(self, n):
    result  = None 
    for i in range(len(self.array) - 1, len(self.array) - 1 - n, -1):  
      result = self.array[i]
    return result 
  
  def nth_largest(self, n):
    buffer = copy.deepcopy(self.array)
    index = None
    
    if n >= len(buffer):
      return None

    for i in range(0, len(buffer) - 1):
      index = 0
      for k in range(0, len(buffer)):
        if buffer[k] < buffer[index] :
          buffer[k], buffer[index] = buffer[index] , buffer[k] 
          index += 1 
    return buffer[len(buffer) - n]      

  def shuffle(self):
    index = 0
    r = None 
    for i in range(0, len(self.array)):
      r = int(math.floor(random.random() * len(self.array))) 
      self.array[index], self.array[r] = self.array[r], self.array[index]
      index += 1
    return self.array
  
  def remove_range(self, start, end):
    index = 0
    while start or end:
      print([start, end, self.array])
      if start:
        self.array = self.array[0 : index] + self.array[index + 1: :] 
        start -= 1 
        end -= 1
      elif end:
        self.array = self.array[0 : end + 1]
        end = 0
    return self.array
  
  
  def print_me(self):
    print(self.array)


def skyline_heights(arr):
  result = [] 
  for val in arr:
    if val > 0 and ((not result)  ^   (val > arr[-1]) ):
      result.append(val)
  return result 


def is_credit_card_valid( collection):
  last_digit = collection[-1] 
  data = 0
  digit = None 
  for i in range ( len(collection) - 2, -1, -1):
    digit = collection[i] 
    if i % 2 == 1:
      digit  *= 2
    if digit > 9:
      digit -= 9
    data +=  digit
  data += last_digit
  return data % 10 
 
def intermediate_sums( arr):
  count = 0
  curr_sum = 0
  index = 0
  
  while index < len(arr):
    curr_sum += arr[index]
    count += 1
    index += 1
    if count == 10:
      count = 0
      arr = arr[0 : index - 1] + [curr_sum] + arr[index: :]
      curr_sum = arr[index]      
      index += 1
  if count:
    arr = arr + [curr_sum]
  return arr 
     
def double_trouble(arr):
  index = 0

  while index < len(arr):
    arr = arr[0: index + 1 ] +    [arr[index]] + arr[index+ 1: :]
    index += 2 
  return arr 

def zip_it (array_a, array_b):
  ptr_a = 0
  ptr_b = 0
  result = []

  while ptr_a < len(array_a) or  ptr_b < len(array_b):

    if ptr_a == len(array_a):
      result += array_b[ptr_b ::]
      break

    elif ptr_b == len(array_b):
      result += array_a[ptr_a ::] 
      break
    
    elif ptr_a == ptr_b:
      result += [array_a[ptr_a]]
      ptr_a += 1

    elif ptr_a > ptr_b:
      result += [array_b[ptr_b]]
      ptr_b += 1

  return result

File: chapter3/java/chapter3.java
package algo;

import java.util.Random;

public class Chapter3 
{
	/* custom array structure */ 
	static class Array_ 
	{
		int maxSize, len;
		int[] collection;
		
		Array_()
		{
			len = 0;
			maxSize = 1;
			collection = new int[maxSize];
		}
		
		Array_ pushFront(int value)
		{
			int [] tmp;
			
			if (len >= maxSize) // array is full!
			{
				maxSize++;
				tmp = new int[maxSize];
				for (int i = 0; i < len; i++)
				{
					tmp[i + 1] = collection[i];
				}
				tmp[0] = value;
				collection = tmp;
			}
			else  
			{
				
				/* shift right*/
				for (int i = maxSize - 1; i > 0; i--)
				{
					collection[i] = collection[i - 1];  
				}
				
				/*add to front*/
				collection[0] = value;
			}
			
			len++;
			return this;
		}
		
		Integer popFront()
		{
			Integer result = null;
			int[] replacementArray;
			
			if (len > 0)
			{
				result = collection[0];
				/* shift left*/
				replacementArray = new int[--len];
				for (int i = 1; i < len; i++)
				{
					replacementArray[i - 1] = collection[i]; 
				}
				collection = replacementArray;
			}
			return result;
		}
		
		boolean insertAt(int index, int value) 
		{
			if (index >= 0 && index < len)
			{
				collection[index] = value;
				return true;
			}
			return false;
		}
		
		Integer removeAt(int index)
		{
			Integer result = null;
			int[] replacementArray;
			int replacementArrayIndex = 0;
			
			if ( len <= 0 || (index < 0 || index >= len) )
			{
				return null;
			}
			
			replacementArray = new int [len - 1];
			
			for (int i = 0; i < len; i++) 
			{
				if (i == index) 
				{
					result = collection[i];
				}
				else 
				{
					replacementArray[replacementArrayIndex++] = collection[i];
				}
			}
			
			len--;
			collection = replacementArray;
			return result;
		}
		
		void swapPairs()
		{
			int tmp;
			for (int i = 0; i <len; i+= 2) 
			{
				tmp = collection[i];
				collection[i] = collection[i + 1];
				collection[i + 1] = tmp;
			}
		}
		
		void minToFront()
		{
			int moveValueToFront;
			int minIndex = 0;
			
			if (len <= 1)
			{
				return;
			}
			
			for (int i = 0; i < len; i++)
			{
				if (collection[i] < collection[minIndex]) 
				{
					minIndex = i;
				}
			}
			
			if (minIndex > 0) 
			{

				moveValueToFront = collection[minIndex];
				/* shift right toString ith index*/
				for (int k = minIndex; k > 0; k--)
				{
					collection[k] = collection[k - 1];
				}
				collection[0] = moveValueToFront;
			}
		}
		
		void reverse () 
		{
			int halfLen, tmp;
			
			if (len <=0) 
			{
				return;
			}
			
			halfLen = len / 2;
			for (int i = 0; i < halfLen; i++) 
			{
				tmp = collection[i];
				collection[i] = collection[len -1 - i];
				collection[len -1 - i] = tmp;
			}
		}
		
		void display()
		{
			System.out.print("[");
			if (len > 0)
			{			
				for (int ele: collection) 
				{
					System.out.print(ele + ",");
				}
			}
			System.out.println("]");
		}
		
		void display(int [] collection)
		{
			System.out.print("[");
			if (len > 0)
			{			
				for (int ele: collection) 
				{
					System.out.print(ele + ",");
				}
			}
			System.out.println("]");
		}
		
	
		
		void rotateArr(int shiftBy) 
		{
			int front, back;
			display();
			
			if (shiftBy < 0)
			{
				shiftBy = ( -1 * shiftBy ) % len;
				
				if (shiftBy == 0) 
				{
					return;
				}
				
				while (shiftBy > 0)
				{
					back = collection[0];
					for (int  i = 1; i < len; i++)
					{
						collection[i - 1] = collection[i];
					}
					collection[len - 1] = back;
					shiftBy--;
				}
				
			}
			else if (shiftBy > 0)
			{
				while (shiftBy > 0)
				{
					front = collection[len  - 1];
					for (int  i = len - 1; i > 0; i--)
					{
						collection[i] = collection[i - 1];
					}
					collection[0] = front;
					shiftBy--;
				}
			}
			display();

		}
		
		void filter(int min, int max) 
		{
			int[] newArray;
			for (int i = 0; i < len; i++) 
			{
				if (collection[i] <= min || collection[i] >= max) 
				{
					// shift down 
					for (int k = i; k < len - 1; k++) 
					{
						collection[k] = collection[k + 1];
					}
					collection[len - 1] = 0;
					len--;
				}
			}
			
			newArray = new int[len];
			for (int i = 0; i < len; i++) 
			{
				newArray[i] = collection[i];
			}
			collection = newArray;
		}
		
		int[] concat(int [] source, int [] arr)
		{
			int[] result = new int[source.length + arr.length];
			int source_index = 0, arr_index = 0;
			boolean srcDataWritable, arrDataWritable;
			
			srcDataWritable = arrDataWritable = true;
			
			/* optimized - 2 write per cycle*/
			while (srcDataWritable || arrDataWritable) 
			{
				srcDataWritable = source_index < source.length;
				arrDataWritable = arr_index < arr.length;
				
				if (srcDataWritable)
				{
					result[source_index] = source[source_index]
							;
					source_index++;
				}
				
				if (arrDataWritable)
				{
					result[result.length - 1 - arr_index] = arr[arr_index];
					arr_index++;
				}
			}
			return result;
		}

		
	}
	
	
	
	int [] skylineHeights(int[] heights)
	{
		Array_ result = new Array_();
		int max = 0;
		for (int i = 0; i < heights.length; i++) 
		{
			if (heights[i] > max)
				
			{
				max = heights[i];
				result.pushFront(max);
			}
		}
		
		result.reverse();
		
		result.display();
		return result.collection;
	}
	
	static Integer secondToLast(int [] arr)
	{
		Integer result =  null;
		if (arr.length > 1 ) 
		{
			return arr[arr.length - 1 - 2];
		}
		return result;
	}
	
	static Integer secondLargest(int [] arr)
	{
		Integer first = null, second = null;
		
		for (int  i = 0; i < arr.length; i++) 
		{
			if (first == null || arr[i] > first )
			{
				if (second != null) 
				{
					second = first;					
				}
				first = arr[i];
			}
			
			else if (second == null || arr[i] > second)
			{
				second = arr[i];
			}
		
		}
		
		if (first != null && second != null)
		{
			return second;
		}
		
		return null;
	}
	
	static Integer nthToLast (int [] arr, int n)
	{
		if (arr.length == 0) 
		{
			return null;
		}
		
		if (n > 1 && n <= arr.length)
		{
			return arr[arr.length - 1 - n];
		}
		return null;
	}
	
	static Integer nthLargest(int[] arr, int n) 
	{
		Integer threshold , currLargest ;
		
		currLargest = threshold = null;
		
		if (arr.length == 0 || n <= 0 || n > arr.length)
		{
			return null;
		}
		
		for (; n > 0; n--)
		{
			currLargest = null;
			for (int ele: arr) 
			{
				if (threshold == null || (ele < threshold && true))  // true required!!
				{
					if (currLargest == null  ||  ele > currLargest) 
					{
						currLargest = ele;
					}
				}
			}
			threshold = currLargest;
		}

		return currLargest;
	}
	
	static boolean  isCreditCardValid(int[] digitArr)
	{
		int lastDigit;
		int sum = 0;
		
		if (digitArr.length == 0) 
		{
			return false;
		}
		
		lastDigit = digitArr[digitArr.length - 1];
		
		for (int i = digitArr.length - 2; i > 0; i-= 2) 
		{
			digitArr[i] *= 2;

			if (digitArr[i] > 9) 
			{
				digitArr[i] -= 9;
			}	
		}
		
		for (int i = 0; i < digitArr.length - 1; i++)
		{
			sum += digitArr[i];
		}
		
		
		sum += lastDigit;
		
		return sum % 10 == 0;
	}
	
	/* swap*/
	static void swap (int[] collection, int posA, int posB)
	{
		int tmp = collection[posA];
		collection[posA] = collection[posB];
		collection[posB] = tmp;
	}
	
	static void shuffle (int[] arr)
	{
		Random random = new Random();
		int walk = 0;
		int rIndex;
		
		while (walk  < arr.length ) 
		{
			rIndex = random.ints(walk , arr.length).findAny().getAsInt();
			swap(arr, walk, rIndex);
			walk++;
		}
	}
	
	static int[] zipIt (int[] a, int[] b)
	{
		int [] result = new int[a.length + b.length];
		int aIndex = 0, bIndex = 0, wrIndex = 0;
		
		while (wrIndex < result.length)
		{
			if (aIndex < a.length)
			{
				result[wrIndex++] = a[aIndex++];
			}
			
			if (bIndex < b.length) 
			{
				result[wrIndex++] = b[bIndex++];
			}
		}
		return result;
	}
	
	public static void main(String args[])
	{
		Array_ arr_ = new Array_();
		for (int i = 10; i > 0; i--)
		{		}
		
		arr_.pushFront(5);
		arr_.pushFront(4);
		arr_.pushFront(3);
		arr_.pushFront(2);
		arr_.pushFront(1);
		
		int[] m = {1, 2};
		int[] n = {10, 20, 30, 40};
		
		for (int ele: zipIt(m, n) )
		{
			System.out.print("[" +  ele + "]");
		}

		
	}
}
